# **NYC Green Taxi Ridership Analysis**

**Team 7**: Chamnan Suon, Anna Baldzikowski, Allison Kramer, Akhil Nair, Mudongfang Li

## **1. Introduction**
### **1.1 Problem Statement**
This project aims to find patterns and trends in New York City Green taxi trips from 2019 to 2023. We will evaluate taxi ridership and columns including passenger count, trip distance, payment type, fare amount, tip amount, and tolls amount in different neighborhoods. These findings will matter to taxi drivers, city officials, and city planners who handle traffic, congestion pricing, and number/timing of taxis on the road. The goals and metrics we will track include:
Analyze ride trends (like passenger count) before, during, and after COVID
Explore links between fares, tips, payment type, and revenue patterns
Examine geographic patterns, and also how they correlate with congestion pricing

### **1.2 Executive Summary**
This project aims to explore various trends in NYC green taxi ridership between 2019 and 2023. The five year interval includes pre-COVID and post-COVID periods, which would allow the studies of different trends, including  focusing on the effects of COVID-19 on ridership, tipping, and overall urban mobility. From our analyses, we discovered that COVID negatively impacted annual revenue in Green taxis rides. Additionally, airport zones and distant areas show much higher average costs, while central Manhattan and Queens remain low-cost due to shorter trip distances. Across all locations, weekday and weekend costs are nearly identical.





### **1.3 Data and Cleaning**
#### **1.3.1 Data**
We obtained our 6 datasets from [NYC Open Data](https://data.cityofnewyork.us/Transportation/2019-Green-Taxi-Trip-Data/q5mz-t52e/about_data). The first 5 datasets records trips in each respective year, 2019, 2020, 2021, 2022, 2023. The sixth dataset contains a monthly aggregate data on pickup and dropoff records in different location in New York City.

##### • **Pickups and Drop-offs Data Dictionary**
Column | Description | API Field Name | Data Type
-------|-------------|----------------|----------
**Metric Month** | The year and month of the report | metric_month | Text
**Industry** | The broad license class used to describe the industry | industry | Text
**Pickup/Dropoff** | Whether the counts reflect pickups or drop-offs | pickup_dropoff | Text
**Location ID** | Taxi zone number | locationid | Number
**Borough** | NYC borough | borough | Text
**Zone** | Taxi zone name | zone | Text
**Trip Count** | Count of trips | trip_count | Number

##### • **Trip Records Data Dictionary**
Column | Description | API Field Name | Data Type
-------|-------------|----------------|--------
**VendorID** | A code indicating the TPEP provider that provided the record. | vendorid | Number
**Ipep_pickup_datetime** | The date and time when the meter was engaged. | Ipep_pickup_datetime | Floating Timestamp
**Ipep_dropoff_datetime** | The date and time when the meter was disengaged. | Ipep_dropoff_datetime | Floating Timestamp
**store_and_fwd_flag** | This flag indicates whether the trip record was held in vehicle memory before sending to the vendor, aka “store and forward,” because the vehicle did not have a connection to the server. | store_and_fwd_flag | Text
**RatecodeID** | The final rate code in effect at the end of the trip. | ratecodeid | Number
**PULocationID** | TLC Taxi Zone in which the taximeter was engaged | pulocationid | Number
**DOLocationID** | TLC Taxi Zone in which the taximeter was disengaged | dolocationid | Number
**passenger_count** | The number of passengers in the vehicle. | passenger_count | Number
**trip_distance** | The elapsed trip distance in miles reported by the taximeter. | trip_distance | Number
**fare_amount** | The time-and-distance fare calculated by the meter. For additional information on the following columns, see https://www.nyc.gov/site/tlc/passengers/taxi-fare.page | fare_amount | Number
**extra** | Miscellaneous extras and surcharges. | extra | Number
**mta_tax** | Tax that is automatically triggered based on the metered rate in use. | mta_tax | Number
**tip_amount** | Tip amount: This field is automatically populated for credit card tips. Cash tips are not included. | tip_amount | Number
**tolls_amount** | Total amount of all tolls paid in trip. | tolls_amount | Number
**ehail_fee** | Currently unused | ehail_fee | Text
**improvement_surcharge** | Improvement surcharge assessed trips at the flag drop. The improvement surcharge began being levied in 2015. | improvement_surcharge | Number
**total_amount** | The total amount charged to passengers. Does not include cash tips. | total_amount | Number
**payment_type** | A numeric code signifying how the passenger paid for the trip. | payment_type | Number
**trip_type** | A code indicating whether the trip was a street-hail or a dispatch that is automatically assigned based on the metered rate in use but can be altered by the driver. | trip_type | Number
**congestion_surcharge** | Total amount collected in trip for NYS congestion surcharge. | congestion_surcharge | Number

#### **1.3.2 Data Clearning**


Our data cleaning process is divided into two parts - for the first part we cleaned 5 years data of trip records and then we cleaned pick and drop-off locations dataset in the second part.

Note that in part 1, since we have 5 datasets, each represents one year of trip records data, there are some repetitive work in the cleaning process. To reduce the lenthy of repetition work, we displayed code cells and result of 2019 dataset cleaning and hided 2020, 2021, 2022, and 2023 as the process are very similar.

##### **Loading Datasets**
Loaded datasets into `775-bigquery-bucket` and converted `lpep_pickup_datetime` and `lpep_dropoff_datetime` columns to string type for consistency and better conversion down the line later.

In [101]:
################## 2019 ###################
%%bigquery
CREATE OR REPLACE EXTERNAL TABLE `ba-775-cs.775_datasets.2019_trip_records`
(
  VendorID INT64,
  lpep_pickup_datetime STRING,
  lpep_dropoff_datetime STRING,
  store_and_fwd_flag STRING,
  RatecodeID INT64,
  PULocationID INT64,
  DOLocationID INT64,
  passenger_count INT64,
  trip_distance FLOAT64,
  fare_amount FLOAT64,
  extra FLOAT64,
  mta_tax FLOAT64,
  tip_amount FLOAT64,
  tolls_amount FLOAT64,
  ehail_fee STRING,
  improvement_surcharge FLOAT64,
  total_amount FLOAT64,
  payment_type INT64,
  trip_type INT64,
  congestion_surcharge FLOAT64
)
OPTIONS (
  format = 'CSV',
  uris = ['gs://775-bigquery-bucket/2019_Green_Taxi_Trips.csv'],
  skip_leading_rows = 1,
  field_delimiter = ','
);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [102]:
# @title
################## 2020 ###################
%%bigquery
CREATE OR REPLACE EXTERNAL TABLE `ba-775-cs.775_datasets.2020_trip_records`
(
  VendorID INT64,
  lpep_pickup_datetime STRING,
  lpep_dropoff_datetime STRING,
  store_and_fwd_flag STRING,
  RatecodeID INT64,
  PULocationID INT64,
  DOLocationID INT64,
  passenger_count INT64,
  trip_distance FLOAT64,
  fare_amount FLOAT64,
  extra FLOAT64,
  mta_tax FLOAT64,
  tip_amount FLOAT64,
  tolls_amount FLOAT64,
  ehail_fee STRING,
  improvement_surcharge FLOAT64,
  total_amount FLOAT64,
  payment_type INT64,
  trip_type INT64,
  congestion_surcharge FLOAT64
)
OPTIONS (
  format = 'CSV',
  uris = ['gs://775-bigquery-bucket/2020_Green_Taxi_Trips.csv'],
  skip_leading_rows = 1,
  field_delimiter = ','
);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [103]:
# @title
################### 2021 #####################
%%bigquery
CREATE OR REPLACE EXTERNAL TABLE `ba-775-cs.775_datasets.2021_trip_records`
(
  VendorID INT64,
  lpep_pickup_datetime STRING,
  lpep_dropoff_datetime STRING,
  store_and_fwd_flag STRING,
  RatecodeID INT64,
  PULocationID INT64,
  DOLocationID INT64,
  passenger_count INT64,
  trip_distance FLOAT64,
  fare_amount FLOAT64,
  extra FLOAT64,
  mta_tax FLOAT64,
  tip_amount FLOAT64,
  tolls_amount FLOAT64,
  ehail_fee STRING,
  improvement_surcharge FLOAT64,
  total_amount FLOAT64,
  payment_type INT64,
  trip_type INT64,
  congestion_surcharge FLOAT64
)
OPTIONS (
  format = 'CSV',
  uris = ['gs://775-bigquery-bucket/2021_Green_Taxi_Trips.csv'],
  skip_leading_rows = 1,
  field_delimiter = ','
);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [104]:
# @title
################### 2022 ######################
%%bigquery
CREATE OR REPLACE EXTERNAL TABLE `ba-775-cs.775_datasets.2022_trip_records`
(
  VendorID INT64,
  lpep_pickup_datetime STRING,
  lpep_dropoff_datetime STRING,
  store_and_fwd_flag STRING,
  RatecodeID INT64,
  PULocationID INT64,
  DOLocationID INT64,
  passenger_count INT64,
  trip_distance FLOAT64,
  fare_amount FLOAT64,
  extra FLOAT64,
  mta_tax FLOAT64,
  tip_amount FLOAT64,
  tolls_amount FLOAT64,
  ehail_fee STRING,
  improvement_surcharge FLOAT64,
  total_amount FLOAT64,
  payment_type INT64,
  trip_type INT64,
  congestion_surcharge FLOAT64
)
OPTIONS (
  format = 'CSV',
  uris = ['gs://775-bigquery-bucket/2022_Green_Taxi_Trips.csv'],
  skip_leading_rows = 1,
  field_delimiter = ','
);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [105]:
# @title
################### 2023 ########################
%%bigquery
CREATE OR REPLACE EXTERNAL TABLE `ba-775-cs.775_datasets.2023_trip_records`
(
  VendorID INT64,
  lpep_pickup_datetime STRING,
  lpep_dropoff_datetime STRING,
  store_and_fwd_flag STRING,
  RatecodeID INT64,
  PULocationID INT64,
  DOLocationID INT64,
  passenger_count INT64,
  trip_distance FLOAT64,
  fare_amount FLOAT64,
  extra FLOAT64,
  mta_tax FLOAT64,
  tip_amount FLOAT64,
  tolls_amount FLOAT64,
  ehail_fee STRING,
  improvement_surcharge FLOAT64,
  total_amount FLOAT64,
  payment_type INT64,
  trip_type INT64,
  congestion_surcharge FLOAT64
)
OPTIONS (
  format = 'CSV',
  uris = ['gs://775-bigquery-bucket/2023_Green_Taxi_Trips.csv'],
  skip_leading_rows = 1,
  field_delimiter = ','
);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

##### **Handling spilled over data and date columns**
Upon data inspection, we learned that there are some old data spilled over to the year after. For instance, 2019 dataset should contain only 2019 data but has data from other year. To prevent duplication issue, we developed SQL query to remove any rows that do not belong in their respective year.

Additionally, changed `lpep_pickup_datetime` and `lpep_dropoff_datetime` date columns from string type to timestamp.

Similarly, we displayed code cell and result of 2019 dataset and hided the rest of the years.

In [106]:
################# 2019 #####################
# VERIFY THAT THERE IS DATA FROM OTHER YEAR IN 2019 DATASET
%%bigquery
SELECT
  EXTRACT(YEAR FROM SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime)) AS year,
  COUNT(*) AS cnt
FROM `ba-775-cs.775_datasets.2019_trip_records`
GROUP BY year
ORDER BY year;

Query is running:   0%|          |

Downloading:   0%|          |

year      cnt
0  2008       16
1  2009       96
2  2010       52
3  2018       67
4  2019  6043777
5  2020       40
6  2035        1
7  2062        1

In [107]:
# FILTERED OUT NON-2019 DATA AND CONVERTED DATE STRING TO TIMESTAMP
%%bigquery
CREATE OR REPLACE TABLE `ba-775-cs.775_datasets.2019_trip_records_date_cleaned` AS
SELECT
  SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime) AS pickup_ts,
  SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_dropoff_datetime) AS dropoff_ts,
  * EXCEPT(lpep_pickup_datetime, lpep_dropoff_datetime)
FROM `ba-775-cs.775_datasets.2019_trip_records`
WHERE EXTRACT(YEAR FROM SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime)) = 2019; --THIS LINE USED TO REMOVE NON-2019 DATA

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [108]:
# CONFIRMED THAT DATE AND OTHER COLUMNS ARE IN THEIR PROPER DATATYPE
%%bigquery
select column_name, data_type
from `775_datasets.INFORMATION_SCHEMA.COLUMNS`
WHERE table_name = '2019_trip_records_date_cleaned';

Query is running:   0%|          |

Downloading:   0%|          |

column_name  data_type
0               pickup_ts  TIMESTAMP
1              dropoff_ts  TIMESTAMP
2                VendorID      INT64
3      store_and_fwd_flag     STRING
4              RatecodeID      INT64
5            PULocationID      INT64
6            DOLocationID      INT64
7         passenger_count      INT64
8           trip_distance    FLOAT64
9             fare_amount    FLOAT64
10                  extra    FLOAT64
11                mta_tax    FLOAT64
12             tip_amount    FLOAT64
13           tolls_amount    FLOAT64
14              ehail_fee     STRING
15  improvement_surcharge    FLOAT64
16           total_amount    FLOAT64
17           payment_type      INT64
18              trip_type      INT64
19   congestion_surcharge    FLOAT64

In [112]:
################# 2020 #####################
# VERIFY THAT THERE IS DATA FROM OTHER YEARS IN 2020 DATASET
%%bigquery
WITH cte AS (
  SELECT
    FORMAT_TIMESTAMP('%m/%d/%Y %I:%M:%S %p',
      SAFE.PARSE_TIMESTAMP('%Y %b %d %I:%M:%S %p', lpep_pickup_datetime)
    ) AS formatted_date
  FROM `ba-775-cs.775_datasets.2020_trip_records`
)
SELECT
  EXTRACT(YEAR FROM PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', formatted_date)) AS year,
  COUNT(*) AS cnt
FROM cte
GROUP BY year
ORDER BY year;


Query is running:   0%|          |

Downloading:   0%|          |

year      cnt
0  2008        4
1  2009       25
2  2010        3
3  2019       19
4  2020  1733998
5  2021        1
6  2041        1

In [113]:
# FILTERED OUT NON-2020 DATA AND CONVERTED DATE STRING TO TIMESTAMP
%%bigquery
CREATE OR REPLACE TABLE `ba-775-cs.775_datasets.2020_trip_records_date_cleaned` AS
WITH cte AS (
  SELECT
    PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', FORMAT_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', SAFE.PARSE_TIMESTAMP('%Y %b %d %I:%M:%S %p', lpep_pickup_datetime))) AS pickup_ts,
    PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', FORMAT_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', SAFE.PARSE_TIMESTAMP('%Y %b %d %I:%M:%S %p', lpep_dropoff_datetime))) AS dropoff_ts,
    *
  FROM `ba-775-cs.775_datasets.2020_trip_records`
)

SELECT
  * EXCEPT(lpep_pickup_datetime, lpep_dropoff_datetime)
FROM cte
WHERE EXTRACT(YEAR FROM pickup_ts) = 2020;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [114]:
# CONFIRMED THAT DATE COLUMNS ARE TIMESTAMP TYPE
%%bigquery
select column_name, data_type
from `775_datasets.INFORMATION_SCHEMA.COLUMNS`
WHERE table_name = '2020_trip_records_date_cleaned';

Query is running:   0%|          |

Downloading:   0%|          |

column_name  data_type
0               pickup_ts  TIMESTAMP
1              dropoff_ts  TIMESTAMP
2                VendorID      INT64
3      store_and_fwd_flag     STRING
4              RatecodeID      INT64
5            PULocationID      INT64
6            DOLocationID      INT64
7         passenger_count      INT64
8           trip_distance    FLOAT64
9             fare_amount    FLOAT64
10                  extra    FLOAT64
11                mta_tax    FLOAT64
12             tip_amount    FLOAT64
13           tolls_amount    FLOAT64
14              ehail_fee     STRING
15  improvement_surcharge    FLOAT64
16           total_amount    FLOAT64
17           payment_type      INT64
18              trip_type      INT64
19   congestion_surcharge    FLOAT64

In [115]:
################### 2021 ###################
# VERIFY THAT THERE IS DATA FROM OTHER YEAR IN 2021 DATASET
%%bigquery
WITH cte AS (
  SELECT
    FORMAT_TIMESTAMP('%m/%d/%Y %I:%M:%S %p',
      PARSE_TIMESTAMP('%Y %b %d %I:%M:%S %p', lpep_pickup_datetime)
    ) AS formatted_date
  FROM `ba-775-cs.775_datasets.2021_trip_records`
)
SELECT
  EXTRACT(YEAR FROM PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', formatted_date)) AS year,
  COUNT(*) AS cnt
FROM cte
GROUP BY year
ORDER BY year;

Query is running:   0%|          |

Downloading:   0%|          |

year      cnt
0  2008        4
1  2009       21
2  2010        4
3  2020        1
4  2021  1068725

In [116]:
# FILTERED OUT NON-2021 DATA AND CONVERTED DATE STRING TO TIMESTAMP
%%bigquery
CREATE OR REPLACE TABLE `ba-775-cs.775_datasets.2021_trip_records_date_cleaned` AS
WITH cte AS (
  SELECT
    PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', FORMAT_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', SAFE.PARSE_TIMESTAMP('%Y %b %d %I:%M:%S %p', lpep_pickup_datetime))) AS pickup_ts,
    PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', FORMAT_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', SAFE.PARSE_TIMESTAMP('%Y %b %d %I:%M:%S %p', lpep_dropoff_datetime))) AS dropoff_ts,
    *
  FROM `ba-775-cs.775_datasets.2021_trip_records`
)

SELECT
  * EXCEPT(lpep_pickup_datetime, lpep_dropoff_datetime)
FROM cte
WHERE EXTRACT(YEAR FROM pickup_ts) = 2021;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [117]:
# CONFIRMED THAT DATE COLUMNS ARE TIMESTAMP TYPE
%%bigquery
select column_name, data_type
from `775_datasets.INFORMATION_SCHEMA.COLUMNS`
WHERE table_name = '2021_trip_records_date_cleaned';

Query is running:   0%|          |

Downloading:   0%|          |

column_name  data_type
0               pickup_ts  TIMESTAMP
1              dropoff_ts  TIMESTAMP
2                VendorID      INT64
3      store_and_fwd_flag     STRING
4              RatecodeID      INT64
5            PULocationID      INT64
6            DOLocationID      INT64
7         passenger_count      INT64
8           trip_distance    FLOAT64
9             fare_amount    FLOAT64
10                  extra    FLOAT64
11                mta_tax    FLOAT64
12             tip_amount    FLOAT64
13           tolls_amount    FLOAT64
14              ehail_fee     STRING
15  improvement_surcharge    FLOAT64
16           total_amount    FLOAT64
17           payment_type      INT64
18              trip_type      INT64
19   congestion_surcharge    FLOAT64

In [141]:
################## 2022 ######################
# VERIFY THAT THERE IS DATA FROM OTHER YEARS IN 2022 DATASET
%%bigquery
SELECT
  EXTRACT(YEAR FROM SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime)) AS year,
  COUNT(*) AS cnt
FROM `ba-775-cs.775_datasets.2022_trip_records`
GROUP BY year
ORDER BY year;

Query is running:   0%|          |

Downloading:   0%|          |

year     cnt
0  2008       2
1  2009       5
2  2021       3
3  2022  840392

In [142]:
# FILTERED OUT NON-2022 DATA AND CONVERTED DATE STRING TO TIMESTAMP
%%bigquery
CREATE OR REPLACE TABLE `ba-775-cs.775_datasets.2022_trip_records_date_cleaned` AS
SELECT
  SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime) AS pickup_ts,
  SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_dropoff_datetime) AS dropoff_ts,
  * EXCEPT(lpep_pickup_datetime, lpep_dropoff_datetime)
FROM `ba-775-cs.775_datasets.2022_trip_records`
WHERE EXTRACT(YEAR FROM SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime)) = 2022;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [143]:
# CONFIRMED THAT DATE COLUMNS ARE TIMESTAMP TYPE
%%bigquery
select column_name, data_type
from `775_datasets.INFORMATION_SCHEMA.COLUMNS`
WHERE table_name = '2022_trip_records_date_cleaned';

Query is running:   0%|          |

Downloading:   0%|          |

column_name  data_type
0               pickup_ts  TIMESTAMP
1              dropoff_ts  TIMESTAMP
2                VendorID      INT64
3      store_and_fwd_flag     STRING
4              RatecodeID      INT64
5            PULocationID      INT64
6            DOLocationID      INT64
7         passenger_count      INT64
8           trip_distance    FLOAT64
9             fare_amount    FLOAT64
10                  extra    FLOAT64
11                mta_tax    FLOAT64
12             tip_amount    FLOAT64
13           tolls_amount    FLOAT64
14              ehail_fee     STRING
15  improvement_surcharge    FLOAT64
16           total_amount    FLOAT64
17           payment_type      INT64
18              trip_type      INT64
19   congestion_surcharge    FLOAT64

In [144]:
################## 2023 ####################
# VERIFY THAT THERE IS DATA FROM OTHER YEARS IN 2023 DATASET
%%bigquery
SELECT
  EXTRACT(YEAR FROM SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime)) AS year,
  COUNT(*) AS cnt
FROM `ba-775-cs.775_datasets.2023_trip_records`
GROUP BY year
ORDER BY year;

Query is running:   0%|          |

Downloading:   0%|          |

year     cnt
0  2008       2
1  2009       3
2  2022       2
3  2023  787053

In [145]:
# FILTERED OUT NON-2023 DATA AND CONVERTED DATE STRING TO TIMESTAMP
%%bigquery
CREATE OR REPLACE TABLE `ba-775-cs.775_datasets.2023_trip_records_date_cleaned` AS
SELECT
  SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime) AS pickup_ts,
  SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_dropoff_datetime) AS dropoff_ts,
  * EXCEPT(lpep_pickup_datetime, lpep_dropoff_datetime)
FROM `ba-775-cs.775_datasets.2023_trip_records`
WHERE EXTRACT(YEAR FROM SAFE.PARSE_TIMESTAMP('%m/%d/%Y %I:%M:%S %p', lpep_pickup_datetime)) = 2023;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [146]:
# CONFIRMED THAT DATE COLUMNS ARE TIMESTAMP TYPE
%%bigquery
select column_name, data_type
from `775_datasets.INFORMATION_SCHEMA.COLUMNS`
WHERE table_name = '2023_trip_records_date_cleaned';

Query is running:   0%|          |

Downloading:   0%|          |

column_name  data_type
0               pickup_ts  TIMESTAMP
1              dropoff_ts  TIMESTAMP
2                VendorID      INT64
3      store_and_fwd_flag     STRING
4              RatecodeID      INT64
5            PULocationID      INT64
6            DOLocationID      INT64
7         passenger_count      INT64
8           trip_distance    FLOAT64
9             fare_amount    FLOAT64
10                  extra    FLOAT64
11                mta_tax    FLOAT64
12             tip_amount    FLOAT64
13           tolls_amount    FLOAT64
14              ehail_fee     STRING
15  improvement_surcharge    FLOAT64
16           total_amount    FLOAT64
17           payment_type      INT64
18              trip_type      INT64
19   congestion_surcharge    FLOAT64

#### **Find and Replace Missing Values**

After finishing filtering out each dataset with non-belonging data, we ran query below to display count of total missing values in each colum in a dataset. We applied this query to the rest of the datasets.

Moreover, we learned that each dataset has missing values in some of the columns except `pickup and dropoff locations` dataset. We handled missing values in each column of each dataset as below:

• Drop `ehail_fee` column as this is an empty column and has been informed in the data dictionary documentation that this column is currently unused

• Drop `store_and_fwd_flag` column as it is not relevant to our analysis

• Replace missing values in `RatecodelD` with 99

• Replace missing values in `passenger_count` with mode

• Replace missing values in `payment_type` with 5

In [ ]:
################## 2019 ###################
%%bigquery
-- COUNT NULL VALUES IN EACH COLUMN
SELECT
  COUNTIF(vendorid IS NULL) AS vendorid_nulls,
  COUNTIF(pickup_ts IS NULL) AS pickup_ts_nulls,
  COUNTIF(dropoff_ts IS NULL) AS dropoff_ts_nulls,
  COUNTIF(store_and_fwd_flag IS NULL) AS store_and_fwd_flag_nulls,
  COUNTIF(RatecodeID IS NULL) AS RatecodeID_nulls,
  COUNTIF(PULocationID IS NULL) AS PULocationID_nulls,
  COUNTIF(DOLocationID IS NULL) AS DOLocationID_nulls,
  COUNTIF(passenger_count IS NULL) AS passenger_count_nulls,
  COUNTIF(trip_distance IS NULL) AS trip_distance_nulls,
  COUNTIF(fare_amount IS NULL) AS fare_amount_nulls,
  COUNTIF(extra IS NULL) AS extra_nulls,
  COUNTIF(mta_tax IS NULL) AS mta_tax_nulls,
  COUNTIF(tip_amount IS NULL) AS tip_amount_nulls,
  COUNTIF(tolls_amount IS NULL) AS tolls_amount_nulls,
  COUNTIF(ehail_fee IS NULL) AS ehail_fee_nulls,
  COUNTIF(improvement_surcharge IS NULL) AS improvement_surcharge_nulls,
  COUNTIF(total_amount IS NULL) AS total_amount_nulls,
  COUNTIF(payment_type IS NULL) AS payment_type_nulls,
  COUNTIF(trip_type IS NULL) AS trip_type_nulls,
  COUNTIF(congestion_surcharge IS NULL) AS congestion_surcharge_nulls
FROM `ba-775-cs.775_datasets.2019_trip_records_date_cleaned`;

Query is running:   0%|          |

Downloading:   0%|          |

vendorid_nulls  pickup_ts_nulls  dropoff_ts_nulls  \
0          414107                0                 0   

   store_and_fwd_flag_nulls  RatecodeID_nulls  PULocationID_nulls  \
0                    414107            414107                   0   

   DOLocationID_nulls  passenger_count_nulls  trip_distance_nulls  \
0                   0                 414107                    0   

   fare_amount_nulls  extra_nulls  mta_tax_nulls  tip_amount_nulls  \
0                  0            0              0                 0   

   tolls_amount_nulls  ehail_fee_nulls  improvement_surcharge_nulls  \
0                   0          6043423                            2   

   total_amount_nulls  payment_type_nulls  trip_type_nulls  \
0                   0              414107           414466   

   congestion_surcharge_nulls  
0                      960401

In [147]:
# HANDLING MISSING VALUES
%%bigquery
CREATE OR REPLACE VIEW
  `ba-775-cs.775_datasets.2019_trip_records_view` AS
WITH
  cte AS (
  SELECT
    passenger_count AS passenger_mode,
    COUNT(*) AS mode
  FROM
    `ba-775-cs.775_datasets.2019_trip_records_date_cleaned`
  WHERE
    passenger_count IS NOT NULL
  GROUP BY
    passenger_count
  ORDER BY
    mode DESC
  LIMIT
    1 )
SELECT
  COALESCE(RatecodeID, 99) AS RatecodeID_imputed, --imputed missing values with 99 as it means 'unknown' in data dictionary
  COALESCE(payment_type, 5) AS payment_type_imputed, --imputed missing values with 5 as it means 'unknown' in data dictionary
  COALESCE(passenger_count, (
    SELECT passenger_mode
    FROM cte
  )) AS passenger_count_imputed, --imputed missing values with mode
  * EXCEPT (ehail_fee, store_and_fwd_flag, RatecodeID, payment_type, passenger_count)
FROM
  `ba-775-cs.775_datasets.2019_trip_records_date_cleaned`;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [ ]:
#################### 2020 ########################
%%bigquery
-- COUNT NULL VALUES IN EACH COLUMN
SELECT
  COUNTIF(vendorid IS NULL) AS vendorid_nulls,
  COUNTIF(pickup_ts IS NULL) AS pickup_ts_nulls,
  COUNTIF(dropoff_ts IS NULL) AS dropoff_ts_nulls,
  COUNTIF(store_and_fwd_flag IS NULL) AS store_and_fwd_flag_nulls,
  COUNTIF(RatecodeID IS NULL) AS RatecodeID_nulls,
  COUNTIF(PULocationID IS NULL) AS PULocationID_nulls,
  COUNTIF(DOLocationID IS NULL) AS DOLocationID_nulls,
  COUNTIF(passenger_count IS NULL) AS passenger_count_nulls,
  COUNTIF(trip_distance IS NULL) AS trip_distance_nulls,
  COUNTIF(fare_amount IS NULL) AS fare_amount_nulls,
  COUNTIF(extra IS NULL) AS extra_nulls,
  COUNTIF(mta_tax IS NULL) AS mta_tax_nulls,
  COUNTIF(tip_amount IS NULL) AS tip_amount_nulls,
  COUNTIF(tolls_amount IS NULL) AS tolls_amount_nulls,
  COUNTIF(ehail_fee IS NULL) AS ehail_fee_nulls,
  COUNTIF(improvement_surcharge IS NULL) AS improvement_surcharge_nulls,
  COUNTIF(total_amount IS NULL) AS total_amount_nulls,
  COUNTIF(payment_type IS NULL) AS payment_type_nulls,
  COUNTIF(trip_type IS NULL) AS trip_type_nulls,
  COUNTIF(congestion_surcharge IS NULL) AS congestion_surcharge_nulls
FROM `ba-775-cs.775_datasets.2020_trip_records_date_cleaned`;

Query is running:   0%|          |

Downloading:   0%|          |

vendorid_nulls  pickup_ts_nulls  dropoff_ts_nulls  \
0          528092                0                 0   

   store_and_fwd_flag_nulls  RatecodeID_nulls  PULocationID_nulls  \
0                    528092            528092                   0   

   DOLocationID_nulls  passenger_count_nulls  trip_distance_nulls  \
0                   0                 528092                    0   

   fare_amount_nulls  extra_nulls  mta_tax_nulls  tip_amount_nulls  \
0                  0            0              0                 0   

   tolls_amount_nulls  ehail_fee_nulls  improvement_surcharge_nulls  \
0                   0          1733998                            0   

   total_amount_nulls  payment_type_nulls  trip_type_nulls  \
0                   0              528092           528097   

   congestion_surcharge_nulls  
0                      528092

In [148]:
%%bigquery
CREATE OR REPLACE VIEW
  `775_datasets.2020_trip_records_view` AS
WITH
  cte AS (
  SELECT
    passenger_count AS passenger_mode,
    COUNT(*) AS mode
  FROM
    `775_datasets.2020_trip_records_date_cleaned`
  WHERE
    passenger_count IS NOT NULL
  GROUP BY
    passenger_count
  ORDER BY
    mode DESC
  LIMIT
    1 )
SELECT
  COALESCE(RatecodeID, 99) AS ratecodeID_imputed, --replace missing values with 99
  COALESCE(payment_type, 5) AS payment_type_imputed, --replace missing values with 5
  COALESCE(passenger_count, ( --replace missing values with mode
    SELECT
      passenger_mode
    FROM
      cte)) AS passenger_count,
  * EXCEPT (ehail_fee,
    store_and_fwd_flag,
    RatecodeID,
    payment_type,
    passenger_count)
FROM
  `775_datasets.2020_trip_records_date_cleaned`

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [149]:
################# 2021 #####################
%%bigquery
-- COUNT NULL VALUES IN EACH COLUMN
SELECT
  COUNTIF(vendorid IS NULL) AS vendorid_nulls,
  COUNTIF(pickup_ts IS NULL) AS pickup_ts_nulls,
  COUNTIF(dropoff_ts IS NULL) AS dropoff_ts_nulls,
  COUNTIF(store_and_fwd_flag IS NULL) AS store_and_fwd_flag_nulls,
  COUNTIF(RatecodeID IS NULL) AS RatecodeID_nulls,
  COUNTIF(PULocationID IS NULL) AS PULocationID_nulls,
  COUNTIF(DOLocationID IS NULL) AS DOLocationID_nulls,
  COUNTIF(passenger_count IS NULL) AS passenger_count_nulls,
  COUNTIF(trip_distance IS NULL) AS trip_distance_nulls,
  COUNTIF(fare_amount IS NULL) AS fare_amount_nulls,
  COUNTIF(extra IS NULL) AS extra_nulls,
  COUNTIF(mta_tax IS NULL) AS mta_tax_nulls,
  COUNTIF(tip_amount IS NULL) AS tip_amount_nulls,
  COUNTIF(tolls_amount IS NULL) AS tolls_amount_nulls,
  COUNTIF(ehail_fee IS NULL) AS ehail_fee_nulls,
  COUNTIF(improvement_surcharge IS NULL) AS improvement_surcharge_nulls,
  COUNTIF(total_amount IS NULL) AS total_amount_nulls,
  COUNTIF(payment_type IS NULL) AS payment_type_nulls,
  COUNTIF(trip_type IS NULL) AS trip_type_nulls,
  COUNTIF(congestion_surcharge IS NULL) AS congestion_surcharge_nulls
FROM `ba-775-cs.775_datasets.2021_trip_records_date_cleaned`;

Query is running:   0%|          |

Downloading:   0%|          |

vendorid_nulls  pickup_ts_nulls  dropoff_ts_nulls  \
0          249115                0                 0   

   store_and_fwd_flag_nulls  RatecodeID_nulls  PULocationID_nulls  \
0                    412434            412434                   0   

   DOLocationID_nulls  passenger_count_nulls  trip_distance_nulls  \
0                   0                 412434                    0   

   fare_amount_nulls  extra_nulls  mta_tax_nulls  tip_amount_nulls  \
0                  0            0              0                 0   

   tolls_amount_nulls  ehail_fee_nulls  improvement_surcharge_nulls  \
0                   0          1068725                            0   

   total_amount_nulls  payment_type_nulls  trip_type_nulls  \
0                   0              412434           412434   

   congestion_surcharge_nulls  
0                      412434

In [150]:
%%bigquery
CREATE OR REPLACE VIEW
  `775_datasets.2021_trip_records_view` AS
WITH
  cte AS (
  SELECT
    passenger_count AS passenger_mode,
    COUNT(*) AS mode
  FROM
    `775_datasets.2021_trip_records_date_cleaned`
  WHERE
    passenger_count IS NOT NULL
  GROUP BY
    passenger_count
  ORDER BY
    mode DESC
  LIMIT
    1 )
SELECT
  COALESCE(RatecodeID, 99) AS ratecodeID_imputed, --replace missing values with 99
  COALESCE(payment_type, 5) AS payment_type_imputed, --replace missing values with 5
  COALESCE(passenger_count, ( --replace missing values with mode
    SELECT
      passenger_mode
    FROM
      cte)) AS passenger_count,
  * EXCEPT (ehail_fee,
    store_and_fwd_flag,
    RatecodeID,
    payment_type,
    passenger_count)
FROM
  `775_datasets.2021_trip_records_date_cleaned`

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [151]:
##################### 2022 #######################
%%bigquery
-- COUNT NULL VALUES IN EACH COLUMN
SELECT
  COUNTIF(vendorid IS NULL) AS vendorid_nulls,
  COUNTIF(pickup_ts IS NULL) AS pickup_ts_nulls,
  COUNTIF(dropoff_ts IS NULL) AS dropoff_ts_nulls,
  COUNTIF(store_and_fwd_flag IS NULL) AS store_and_fwd_flag_nulls,
  COUNTIF(RatecodeID IS NULL) AS RatecodeID_nulls,
  COUNTIF(PULocationID IS NULL) AS PULocationID_nulls,
  COUNTIF(DOLocationID IS NULL) AS DOLocationID_nulls,
  COUNTIF(passenger_count IS NULL) AS passenger_count_nulls,
  COUNTIF(trip_distance IS NULL) AS trip_distance_nulls,
  COUNTIF(fare_amount IS NULL) AS fare_amount_nulls,
  COUNTIF(extra IS NULL) AS extra_nulls,
  COUNTIF(mta_tax IS NULL) AS mta_tax_nulls,
  COUNTIF(tip_amount IS NULL) AS tip_amount_nulls,
  COUNTIF(tolls_amount IS NULL) AS tolls_amount_nulls,
  COUNTIF(ehail_fee IS NULL) AS ehail_fee_nulls,
  COUNTIF(improvement_surcharge IS NULL) AS improvement_surcharge_nulls,
  COUNTIF(total_amount IS NULL) AS total_amount_nulls,
  COUNTIF(payment_type IS NULL) AS payment_type_nulls,
  COUNTIF(trip_type IS NULL) AS trip_type_nulls,
  COUNTIF(congestion_surcharge IS NULL) AS congestion_surcharge_nulls
FROM `ba-775-cs.775_datasets.2022_trip_records_date_cleaned`;

Query is running:   0%|          |

Downloading:   0%|          |

vendorid_nulls  pickup_ts_nulls  dropoff_ts_nulls  \
0               0                0                 0   

   store_and_fwd_flag_nulls  RatecodeID_nulls  PULocationID_nulls  \
0                     90340             90340                   0   

   DOLocationID_nulls  passenger_count_nulls  trip_distance_nulls  \
0                   0                  90340                    0   

   fare_amount_nulls  extra_nulls  mta_tax_nulls  tip_amount_nulls  \
0                  0            0              0                 0   

   tolls_amount_nulls  ehail_fee_nulls  improvement_surcharge_nulls  \
0                   0           840392                            0   

   total_amount_nulls  payment_type_nulls  trip_type_nulls  \
0                   0               90340            90355   

   congestion_surcharge_nulls  
0                       90340

In [152]:
%%bigquery
CREATE OR REPLACE VIEW
  `775_datasets.2022_trip_records_view` AS
WITH
  cte AS (
  SELECT
    passenger_count AS passenger_mode,
    COUNT(*) AS mode
  FROM
    `775_datasets.2022_trip_records_date_cleaned`
  WHERE
    passenger_count IS NOT NULL
  GROUP BY
    passenger_count
  ORDER BY
    mode DESC
  LIMIT
    1 )
SELECT
  COALESCE(RatecodeID, 99) AS ratecodeID_imputed, --replace missing values with 99
  COALESCE(payment_type, 5) AS payment_type_imputed, --replace missing values with 5
  COALESCE(passenger_count, ( --replace missing values with mode
    SELECT
      passenger_mode
    FROM
      cte)) AS passenger_count,
  * EXCEPT (ehail_fee,
    store_and_fwd_flag,
    RatecodeID,
    payment_type,
    passenger_count)
FROM
  `775_datasets.2022_trip_records_date_cleaned`

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [153]:
# @title
##################### 2023 ######################
%%bigquery
-- COUNT NULL VALUES IN EACH COLUMN
SELECT
  COUNTIF(vendorid IS NULL) AS vendorid_nulls,
  COUNTIF(pickup_ts IS NULL) AS pickup_ts_nulls,
  COUNTIF(dropoff_ts IS NULL) AS dropoff_ts_nulls,
  COUNTIF(store_and_fwd_flag IS NULL) AS store_and_fwd_flag_nulls,
  COUNTIF(RatecodeID IS NULL) AS RatecodeID_nulls,
  COUNTIF(PULocationID IS NULL) AS PULocationID_nulls,
  COUNTIF(DOLocationID IS NULL) AS DOLocationID_nulls,
  COUNTIF(passenger_count IS NULL) AS passenger_count_nulls,
  COUNTIF(trip_distance IS NULL) AS trip_distance_nulls,
  COUNTIF(fare_amount IS NULL) AS fare_amount_nulls,
  COUNTIF(extra IS NULL) AS extra_nulls,
  COUNTIF(mta_tax IS NULL) AS mta_tax_nulls,
  COUNTIF(tip_amount IS NULL) AS tip_amount_nulls,
  COUNTIF(tolls_amount IS NULL) AS tolls_amount_nulls,
  COUNTIF(ehail_fee IS NULL) AS ehail_fee_nulls,
  COUNTIF(improvement_surcharge IS NULL) AS improvement_surcharge_nulls,
  COUNTIF(total_amount IS NULL) AS total_amount_nulls,
  COUNTIF(payment_type IS NULL) AS payment_type_nulls,
  COUNTIF(trip_type IS NULL) AS trip_type_nulls,
  COUNTIF(congestion_surcharge IS NULL) AS congestion_surcharge_nulls
FROM `ba-775-cs.775_datasets.2023_trip_records_date_cleaned`;

Query is running:   0%|          |

Downloading:   0%|          |

vendorid_nulls  pickup_ts_nulls  dropoff_ts_nulls  \
0               0                0                 0   

   store_and_fwd_flag_nulls  RatecodeID_nulls  PULocationID_nulls  \
0                     55613             55613                   0   

   DOLocationID_nulls  passenger_count_nulls  trip_distance_nulls  \
0                   0                  55613                    0   

   fare_amount_nulls  extra_nulls  mta_tax_nulls  tip_amount_nulls  \
0                  0            0              0                 0   

   tolls_amount_nulls  ehail_fee_nulls  improvement_surcharge_nulls  \
0                   0           787053                            0   

   total_amount_nulls  payment_type_nulls  trip_type_nulls  \
0                   0               55613            55658   

   congestion_surcharge_nulls  
0                       55613

In [154]:
%%bigquery
CREATE OR REPLACE VIEW
  `775_datasets.2023_trip_records_view` AS
WITH
  cte AS (
  SELECT
    passenger_count AS passenger_mode,
    COUNT(*) AS mode
  FROM
    `775_datasets.2023_trip_records_date_cleaned`
  WHERE
    passenger_count IS NOT NULL
  GROUP BY
    passenger_count
  ORDER BY
    mode DESC
  LIMIT
    1 )
SELECT
  COALESCE(RatecodeID, 99) AS ratecodeID_imputed, --replace missing values with 99
  COALESCE(payment_type, 5) AS payment_type_imputed, --replace missing values with 5
  COALESCE(passenger_count, ( --replace missing values with mode
    SELECT
      passenger_mode
    FROM
      cte)) AS passenger_count,
  * EXCEPT (ehail_fee,
    store_and_fwd_flag,
    RatecodeID,
    payment_type,
    passenger_count)
FROM
  `775_datasets.2023_trip_records_date_cleaned`

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

**Combined 2019, 2020, 2021, 2022, 2023 datasets** into view and made available in BigQuery datasets for analysis.

In [155]:
%%bigquery
CREATE OR REPLACE VIEW `ba-775-cs.775_datasets.all_trip_records_view` AS
SELECT * FROM `ba-775-cs.775_datasets.2019_trip_records_view`
UNION DISTINCT
SELECT * FROM `ba-775-cs.775_datasets.2020_trip_records_view`
UNION DISTINCT
SELECT * FROM `ba-775-cs.775_datasets.2021_trip_records_view`
UNION DISTINCT
SELECT * FROM `ba-775-cs.775_datasets.2022_trip_records_view`
UNION DISTINCT
SELECT * FROM `ba-775-cs.775_datasets.2023_trip_records_view`;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

#### **Pickup-Dropoff Locations Dataset**

As discussed, after finishing cleaning five years of trip records datasets, we cleaned our `pickup-dropoff locations` dataset.

Through the cleaning process, we confirmed that there is no missing data in this dataset. For consistency, we converted `metric_month` column to timestamp type. We

Steps performed to clean the dataset:
- loaded dataset into bucket
- reviewed overview of dataset
- counted missing values in each column
- filtered out non-Green taxi data and created view
- confirmed columns datatypes

In [156]:
# Loaded dataset into bucket
%%bigquery
CREATE OR REPLACE EXTERNAL TABLE `ba-775-cs.775_datasets.pickups_dropoffs_monthly`
(
  metric_month STRING,
  industry STRING,
  pickup_dropoff STRING,
  location_id INT64,
  borough STRING,
  zone_location STRING,
  trip_count INT64
)
OPTIONS (
  format = 'CSV',
  uris = ['gs://775-bigquery-bucket/Pickups_and_Drop-offs.csv'],
  skip_leading_rows = 1,
  field_delimiter = ','
);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [157]:
# Overview of the dataset
%%bigquery
select *
from `ba-775-cs.775_datasets.pickups_dropoffs_monthly`
limit 5;

Query is running:   0%|          |

Downloading:   0%|          |

metric_month           industry pickup_dropoff  location_id borough  \
0      2025-09  FHV - High Volume       Drop-off            1     EWR   
1      2025-09  FHV - High Volume       Drop-off            2  Queens   
2      2025-09  FHV - High Volume         Pickup            2  Queens   
3      2025-09  FHV - High Volume       Drop-off            3   Bronx   
4      2025-09  FHV - High Volume         Pickup            3   Bronx   

             zone_location  trip_count  
0           Newark Airport      147894  
1              Jamaica Bay          87  
2              Jamaica Bay          85  
3  Allerton/Pelham Gardens       40225  
4  Allerton/Pelham Gardens       41683

In [158]:
# Counted any missing values in each column of the dataset
%%bigquery
-- COUNT NULL VALUES IN EACH COLUMN
SELECT
  COUNTIF(metric_month IS NULL) AS metric_month_nulls,
  COUNTIF(industry IS NULL) AS industry_nulls,
  COUNTIF(pickup_dropoff IS NULL) AS pickup_dropoff_nulls,
  COUNTIF(location_id IS NULL) AS location_id_nulls,
  COUNTIF(borough IS NULL) AS borough_nulls,
  COUNTIF(zone_location IS NULL) AS zone_location_nulls,
  COUNTIF(trip_count IS NULL) AS trip_count_nulls,
FROM `ba-775-cs.775_datasets.pickups_dropoffs_monthly`;

Query is running:   0%|          |

Downloading:   0%|          |

metric_month_nulls  industry_nulls  pickup_dropoff_nulls  \
0                   0               0                     0   

   location_id_nulls  borough_nulls  zone_location_nulls  trip_count_nulls  
0                  0              0                    0                 0

In [159]:
# Created view to store pickup-dropoff locations dataset
%%bigquery
CREATE OR REPLACE TABLE `ba-775-cs.775_datasets.pickups_dropoffs_monthly_view` AS
SELECT
  TIMESTAMP(CONCAT(metric_month, '-01')) AS metric_month_ts,
  * EXCEPT(metric_month)
FROM `ba-775-cs.775_datasets.pickups_dropoffs_monthly`
WHERE industry = 'Green Cab'
;


Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [160]:
# Confirmed that metric_month_ts has timestamp type
%%bigquery
select column_name, data_type
from `775_datasets.INFORMATION_SCHEMA.COLUMNS`
WHERE table_name = 'pickups_dropoffs_monthly_view';

Query is running:   0%|          |

Downloading:   0%|          |

column_name  data_type
0  metric_month_ts  TIMESTAMP
1         industry     STRING
2   pickup_dropoff     STRING
3      location_id      INT64
4          borough     STRING
5    zone_location     STRING
6       trip_count      INT64

## **2. Exploratory Data Analysis**

###### **Trips Insights**

The following query looks at **total trips per year**.

In [161]:
%%bigquery
--Trips Per Year
SELECT
 EXTRACT(YEAR FROM pickup_ts) AS year,
 COUNT(*) AS total_trips
FROM
 `ba-775-cs.775_datasets.all_trip_records_view`
WHERE
 EXTRACT(YEAR FROM pickup_ts) BETWEEN 2019 AND 2023
GROUP BY
 year
ORDER BY
 year;

Query is running:   0%|          |

Downloading:   0%|          |

year  total_trips
0  2019      6043716
1  2020      1733860
2  2021      1068725
3  2022       840392
4  2023       787053

**Findings:**

The total number of green taxi trips has decreased each and every year from 2019 to 2023. From 6,043,716 to 787,053, that is a decline of 5,256,663 in 5 years. This fits with the context of COVID, as well as the increased popularity of apps like Uber and Lyft. We also know that green taxis have been gradually phased out, beginning in 2023 ([Source](https://www.nbcnewyork.com/news/local/green-cabs-are-being-phased-out-heres-what-will-replace-them/4302496/)).

The following query looks at **total count of trips by day of week** from 2019 to 2023

In [126]:
%%bigquery
-- Total count of trips by day of week (from 2019–2023)
-- %A is shortcut to represent full weekday name (like "Monday")
SELECT
 FORMAT_TIMESTAMP('%A', pickup_ts) AS day_of_week,
 COUNT(*) AS trip_count
FROM
 `ba-775-cs.775_datasets.all_trip_records_view`
WHERE
 EXTRACT(YEAR FROM pickup_ts) BETWEEN 2019 AND 2023
GROUP BY
 day_of_week
ORDER BY
 trip_count DESC;

Query is running:   0%|          |

Downloading:   0%|          |

day_of_week  trip_count
0      Friday     1667488
1    Thursday     1617298
2   Wednesday     1576992
3     Tuesday     1508615
4    Saturday     1474268
5      Monday     1410470
6      Sunday     1218615

**Findings:**

From 2019 to 2021, Friday was the most popular day for green taxi trips. However, in 2022 and 2023, Thursday became the highest day. For all 5 years, Sunday has been the least popular day.

###### **Geo-location Insights**
The following query looks at the **borough with the most trips** between 2019 and 2023.

In [127]:
%%bigquery

SELECT borough, SUM(trip_count) AS total_trips
FROM `ba-775-cs.775_datasets.pickups_dropoffs_monthly_view`
GROUP BY borough
ORDER BY total_trips DESC

Query is running:   0%|          |

Downloading:   0%|          |

borough  total_trips
0      Manhattan      9443167
1         Queens      5860636
2       Brooklyn      5161533
3          Bronx      1714803
4        Unknown        98760
5  Staten Island        17069
6            EWR         4632

**Findings:**

Manhattan has the most number of trips with 9443167 over this period, with the next closest being Queens at 5860636. Which is in line with it's image of being a tourist hub that attracts millions of people to visit every year ([Source](https://timessquarebillboard.com/blog/nyc-tourism-statistics-2025-times-square-by-the-numbers/#:~:text=NYC%20Tourism%20by%20the%20Numbers)).
The lowest number of trips are from the Newark International Airport with 4632 trips occuring in this time-frame.

The following queries shows **top 3 pickup and top 3 dropoff locations**

In [128]:
%%bigquery
SELECT
 SUM(trip_count) AS total_trip_count_annually,
 borough,
 zone_location,
 EXTRACT(YEAR
 FROM
   metric_month_ts) AS year
FROM
 `ba-775-cs.775_datasets.pickups_dropoffs_monthly_view`
WHERE
 EXTRACT(YEAR
 FROM
   metric_month_ts) = 2023
 AND pickup_dropoff = 'Pickup'
GROUP BY
 year,
 borough,
 zone_location
ORDER BY
 1 DESC
LIMIT 3;


Query is running:   0%|          |

Downloading:   0%|          |

total_trip_count_annually    borough      zone_location  year
0                     159158  Manhattan  East Harlem North  2023
1                     101217  Manhattan  East Harlem South  2023
2                      42880  Manhattan     Central Harlem  2023

In [129]:
%%bigquery
SELECT
 SUM(trip_count) AS total_trip_count_annually,
 borough,
 zone_location,
 EXTRACT(YEAR
 FROM
   metric_month_ts) AS year
FROM
 `ba-775-cs.775_datasets.pickups_dropoffs_monthly_view`
WHERE
 EXTRACT(YEAR
 FROM
   metric_month_ts) = 2023
 AND pickup_dropoff = 'Drop-off'
GROUP BY
 year,
 borough,
 zone_location
ORDER BY
 1 DESC
LIMIT
 3;

Query is running:   0%|          |

Downloading:   0%|          |

total_trip_count_annually    borough          zone_location  year
0                      39957  Manhattan      East Harlem South  2023
1                      38665  Manhattan      East Harlem North  2023
2                      31949  Manhattan  Upper East Side North  2023

**Findings**

As shown in queries above, in 2023, top 3 pickup locations are `East Harlem Nort`, `East Harlem South`, and `Central Harlen`. Similarly, top 3 dropoff locations are `East Harlem South`, `East Harlem Nort` and `Upper East Side North`. All these zone locations are located in Manhattan borough, suggesting high traffic area in the New York State.

###### **Date and Travel Time Insights**
The next query looks at the **length of the trip by the time of day** in order to figure out which is the busiest.

In [130]:
%%bigquery
--Query to find the longest average trip time by time of Day
WITH trip_data AS (
  SELECT
    CASE
      WHEN EXTRACT(HOUR FROM pickup_ts) BETWEEN 0 AND 5 THEN 'Early Morning'
      WHEN EXTRACT(HOUR FROM pickup_ts) BETWEEN 6 AND 11 THEN 'Morning'
      WHEN EXTRACT(HOUR FROM pickup_ts) BETWEEN 12 AND 17 THEN 'Afternoon'
      ELSE 'Evening'
    END AS time_of_day,
    TIMESTAMP_DIFF(dropoff_ts, pickup_ts, MINUTE) AS trip_length_minutes
  FROM `ba-775-cs.775_datasets.all_trip_records_view`
  WHERE pickup_ts IS NOT NULL
    AND dropoff_ts IS NOT NULL
    AND dropoff_ts > pickup_ts
)
SELECT
  time_of_day,
  AVG(trip_length_minutes) AS avg_trip_length_minutes,
  COUNT(*) AS total_trips
FROM trip_data
GROUP BY time_of_day
ORDER BY
  CASE time_of_day
    WHEN 'Early Morning' THEN 1
    WHEN 'Morning' THEN 2
    WHEN 'Afternoon' THEN 3
    WHEN 'Evening' THEN 4
  END;

Query is running:   0%|          |

Downloading:   0%|          |

time_of_day  avg_trip_length_minutes  total_trips
0  Early Morning                21.068122       762175
1        Morning                21.675613      2872836
2      Afternoon                22.607137      3895364
3        Evening                18.593065      2923941

**Findings:**

For the above code, we have set ‘buckets’ wherein we have classified the time at which the trips have occurred. The split being:


*   Early Morning- Midnight to 5 am
*   Morning - 6am to 11am
* Afternoon 12pm to 5 pm and
* Evening being the rest of the

From the results of the query, the afternoon seems to be the busiest time of day with a value of 22.61 minutes, this could be a result of more people such as schools getting over, tourists exploring the city and people leaving from their offices to ironically beat the ‘rush hour’ back home. This is supported with the fact that the most number of trips occur in the afternoon, which is a value of 3895364. To give some perspective, the lowest is the early morning, having only 762175 trips during the 5 years. For Early morning and morning the values are quite close to one another 21.07 and 21.67 respectively. The period when cabs take the shortest time to reach their destination is the Evening with trips taking around 18.59 minutes which is surprising since the total number of trips is second highest, yet ~3 minutes faster than Morning which has ~50,000 less trips.


The following query looks at the **average trip length in minutes** grouped by the pickup location ID.

In [131]:
%%bigquery
-- Query to find the average trip length by the Pickup location
WITH trip_lengths AS (
  SELECT
    PULocationID,
    AVG(TIMESTAMP_DIFF(dropoff_ts, pickup_ts, MINUTE)) AS avg_trip_length_min,
    COUNT(*) AS total_trips
  FROM `ba-775-cs.775_datasets.all_trip_records_view`
  WHERE pickup_ts IS NOT NULL
    AND dropoff_ts IS NOT NULL
    AND dropoff_ts > pickup_ts
  GROUP BY PULocationID
),
zones AS (
  SELECT DISTINCT
    location_id,
    borough,
    zone_location
  FROM `ba-775-cs.775_datasets.pickups_dropoffs_monthly_view`
)
SELECT
  t.PULocationID,
  z.borough,
  z.zone_location,
  t.avg_trip_length_min,
  t.total_trips
FROM trip_lengths t
LEFT JOIN zones z
  ON t.PULocationID = z.location_id
ORDER BY t.avg_trip_length_min DESC;


Query is running:   0%|          |

Downloading:   0%|          |

PULocationID        borough  \
0             156  Staten Island   
1             245  Staten Island   
2              44  Staten Island   
3              23  Staten Island   
4               5  Staten Island   
..            ...            ...   
257            41      Manhattan   
258             1            EWR   
259           111       Brooklyn   
260           264        Unknown   
261           105      Manhattan   

                                     zone_location  avg_trip_length_min  \
0                                  Mariners Harbor            61.143227   
1                                    West Brighton            60.679245   
2                           Charleston/Tottenville            59.430168   
3                          Bloomfield/Emerson Hill            59.303051   
4                                    Arden Heights            57.316327   
..                                             ...                  ...   
257                                 Central Harlem            15.638326   
258                                 Newark Airport            15.548523   
259                            Green-Wood Cemetery            15.362445   
260                                             NV             7.696113   
261  Governor's Island/Ellis Island/Liberty Island             1.000000   

     total_trips  
0           1543  
1            106  
2            179  
3           1475  
4            294  
..           ...  
257       597413  
258          237  
259          229  
260        14252  
261            1  

[262 rows x 5 columns]

**Findings:**

It also joins the two tables together on the location in order to find out which specific boroughs and zones have the longest average trip time. From the results of the query, we can see that the borough that had the longest time was Staten Island which occupied the top 9 spots in terms of average trip length. The longest time on average is 61.14 minutes (slightly over an hour of average travel time) from Mariners Harbor in Staten Island.
This can be attributed to its location and how far it is from other parts of New York City. On the other end of the spectrum is Brooklyn; zone location being Green-Wood Cemetery, where the quickest average time is 15.36 minutes. 8 of the 10 lowest time taking trips on average were in Manhattan.


The next few cells will look at the **busiest days for each year**.

In [132]:
%%bigquery
--2019 output
SELECT
 FORMAT_TIMESTAMP('%A', pickup_ts) AS day_of_week,
 COUNT(*) AS trip_count
FROM
 `ba-775-cs.775_datasets.all_trip_records_view`
WHERE
 EXTRACT(YEAR FROM pickup_ts) = 2019
GROUP BY
 day_of_week
ORDER BY
 trip_count DESC;


Query is running:   0%|          |

Downloading:   0%|          |

day_of_week  trip_count
0      Friday      959768
1    Thursday      913347
2    Saturday      896237
3   Wednesday      883261
4     Tuesday      856940
5      Monday      792739
6      Sunday      741424

In [ ]:
# 2020

In [162]:
%%bigquery
SELECT
 FORMAT_TIMESTAMP('%A', pickup_ts) AS day_of_week,
 COUNT(*) AS trip_count
FROM
 `ba-775-cs.775_datasets.all_trip_records_view`
WHERE
 EXTRACT(YEAR FROM pickup_ts) = 2020
GROUP BY
 day_of_week
ORDER BY
 trip_count DESC;


Query is running:   0%|          |

Downloading:   0%|          |

day_of_week  trip_count
0      Friday      282815
1    Thursday      277510
2   Wednesday      272868
3     Tuesday      252986
4      Monday      239377
5    Saturday      224837
6      Sunday      183467

In [ ]:
# 2021

In [163]:
# @title
%%bigquery
SELECT
 FORMAT_TIMESTAMP('%A', pickup_ts) AS day_of_week,
 COUNT(*) AS trip_count
FROM
 `ba-775-cs.775_datasets.all_trip_records_view`
WHERE
 EXTRACT(YEAR FROM pickup_ts) = 2021
GROUP BY
 day_of_week
ORDER BY
 trip_count DESC;


Query is running:   0%|          |

Downloading:   0%|          |

day_of_week  trip_count
0      Friday      171572
1   Wednesday      169625
2    Thursday      169037
3     Tuesday      160071
4      Monday      153164
5    Saturday      138053
6      Sunday      107203

In [ ]:
# 2022

In [164]:
%%bigquery
SELECT
FORMAT_TIMESTAMP('%A', pickup_ts) AS day_of_week,
COUNT(*) AS trip_count
FROM
`ba-775-cs.775_datasets.all_trip_records_view`
WHERE
EXTRACT(YEAR FROM pickup_ts) = 2022
GROUP BY
day_of_week
ORDER BY
trip_count DESC;

Query is running:   0%|          |

Downloading:   0%|          |

day_of_week  trip_count
0    Thursday      132506
1      Friday      131124
2   Wednesday      128879
3     Tuesday      123275
4      Monday      116318
5    Saturday      113337
6      Sunday       94953

In [ ]:
# 2023

In [165]:
%%bigquery
SELECT
FORMAT_TIMESTAMP('%A', pickup_ts) AS day_of_week,
COUNT(*) AS trip_count
FROM
`ba-775-cs.775_datasets.all_trip_records_view`
WHERE
EXTRACT(YEAR FROM pickup_ts) = 2023
GROUP BY
day_of_week
ORDER BY
trip_count DESC;

Query is running:   0%|          |

Downloading:   0%|          |

day_of_week  trip_count
0    Thursday      124898
1   Wednesday      122359
2      Friday      122209
3     Tuesday      115343
4      Monday      108872
5    Saturday      101804
6      Sunday       91568

###### **Revenue Insights**
The following query displays **total revenue per year**

In [137]:
%%bigquery
SELECT
 ROUND(SUM(total_amount)) AS total_revenue,
 EXTRACT(YEAR
 FROM
   pickup_ts) AS year
FROM
 `ba-775-cs.775_datasets.all_trip_records_view`
GROUP BY
 2
ORDER BY
 2;

Query is running:   0%|          |

Downloading:   0%|          |

total_revenue  year
0    104322782.0  2019
1     34959573.0  2020
2     25577692.0  2021
3     16238282.0  2022
4     18776526.0  2023

**Findings:**

NYC Green taxi generated the highest total revenue in 2019 while 2022 has the lowest number. This result displays the dramatic decrease in revenue from over \$104 millions down to \$34 millions between 2019 and 2020, suggesting the negative impact of COVID on the NYC Green taxi trip. Additionally, we witness gradual decrease in revenue even during post-COVID. This may be due to the announcement of Green taxi being phased out, subsequently reducing the number of Green taxis.

The following query shows **average cost of taxi by location**

In [166]:
%%bigquery
WITH location_cost AS (
  SELECT
  DATE(pickup_ts) AS pickup_date,
  PULocationID AS location_id,
  AVG(total_amount) AS avg_total_cost,
  COUNT(PULocationID) AS trip_count

FROM `ba-775-cs.775_datasets.all_trip_records_view`
WHERE pickup_ts IS NOT NULL
AND dropoff_ts IS NOT NULL
AND dropoff_ts > pickup_ts
GROUP BY location_id, pickup_date
),

zones AS (
SELECT DISTINCT
        location_id,
        zone_location,
        borough
    FROM `ba-775-cs.775_datasets.pickups_dropoffs_monthly`
)
SELECT
l.pickup_date,
l.location_id,
z.borough,
z.zone_location,
l.avg_total_cost,
l.trip_count,


FROM location_cost l
LEFT JOIN zones z
  ON l.location_id = z.location_id

WHERE
  z.zone_location IS NOT NULL
  AND z.borough IS NOT NULL
  and extract(year from l.pickup_date) = 2019


ORDER BY
  avg_total_cost DESC
  LIMIT 4;

Query is running:   0%|          |

Downloading:   0%|          |

pickup_date  location_id        borough                zone_location  \
0  2019-09-26          265        Unknown                           NA   
1  2019-01-10          118  Staten Island  Heartland Village/Todt Hill   
2  2019-10-04          132         Queens                  JFK Airport   
3  2019-11-17          180         Queens                   Ozone Park   

   avg_total_cost  trip_count  
0      305.772857           7  
1      300.000000           1  
2      264.000000           1  
3      252.500000           2

**Findings：**

Airport locations are the most expensive. Places like JFK Airport and Newark Airport have the highest average costs because these trips are long and include tolls/surcharges.
Staten Island locations are also high-cost. These neighborhoods are far from most destinations, so the average trip cost is higher.
Manhattan and central Queens are much cheaper. These areas have many short, local trips, so average costs are usually $20–40.
Locations with very few trips can show “artificially high” averages.If a location has only 1–5 trips, its average cost can be easily distorted by one expensive ride.

The following query shows **average cost of taxi by weekend VS weekday**, attempting to find out if weekend trips are more expensive than weekday trips.

In [167]:
%%bigquery
-- Compute average taxi cost for Weekend vs Weekday (Green Cab only)

SELECT
  CASE
    WHEN EXTRACT(DAYOFWEEK FROM t.pickup_ts) IN (1, 7) THEN 'Weekend'   -- Sunday=1, Saturday=7
    ELSE 'Weekday'
  END AS day_type,                    -- Classify each trip as Weekend / Weekday

  AVG(total_amount) AS avg_total_cost,  -- Total trip cost
  COUNT(*) AS trip_count              -- Number of trips

FROM `ba-775-cs.775_datasets.all_trip_records_view` AS t

-- Join lookup table, filtered to Green Cab pickup zones
LEFT JOIN (
    SELECT *
    FROM `ba-775-cs.775_datasets.pickups_dropoffs_monthly`
    WHERE pickup_dropoff = 'Pickup'    -- Pickup locations only related to costs
) AS z
ON t.PULocationID = z.location_id

GROUP BY
  day_type

ORDER BY
  avg_total_cost DESC;

Query is running:   0%|          |

Downloading:   0%|          |

day_type  avg_total_cost  trip_count
0  Weekday       19.286492  1861266595
1  Weekend       18.352436   644279442

**Findings：**

Based on Green Cab data, the average trip cost is very similar between weekdays and weekends. There is no major price difference between weekday and weekend trips. This suggests that Green Cab pricing is mainly driven by trip distance and location, not by the day of the week.


## **3. Conclusion**

From 2019 to 2023, NYC green taxi activity declined sharply, with total trips dropping from over 6 million to fewer than 800,000—a loss of more than 5 million rides. This downward trend aligns with major disruptions such as COVID-19, the rapid rise of Uber and Lyft, and the official phase-out of the green taxi program.

Throughout the five-year period, Manhattan consistently generated the highest demand, with East Harlem and nearby neighborhoods ranking as the top pickup and dropoff locations. In contrast, Newark Airport saw the fewest trips. Time-of-day patterns reveal that afternoons are the busiest and generally the slowest periods, while evenings—despite high volume—have the shortest average travel times.

Borough-level results show that Staten Island trips are the longest on average due to distance, while many of the shortest trips occur in Manhattan and Brooklyn. Financially, revenue peaked in 2019 and then fell steeply during the pandemic, continuing to decline in later years as green taxi operations scaled back.

Cost analysis indicates that airport and Staten Island trips are the most expensive, while Manhattan and central Queens trips are cheaper due to shorter distances. Weekdays and weekends show similar average costs, suggesting that pricing depends more on distance than day of week.

## **4. Limitation**

The phase-out of green taxis and large causes drop in trip volume—especially after 2020, which may distort trends and make later-year results less representative of typical rider behavior. Additionally, zones with very few trips also create unreliable averages for cost and duration.

## **5. Next Steps**

- Further analyze high-demand zones (e.g., East Harlem) and low-volume zones (e.g., Newark Airport) to understand differences in demand, duration, and cost.

- Examine how afternoon congestion and evening travel efficiency vary across boroughs to better explain duration differences within the existing groups.

- Compare cost, duration, and trip volume patterns across years to understand how service shifted alongside the decline in trips.

## **6. Generative AI Statement**:
During the development of this project, we used Generative AI tools such as ChatGPT to support specific tasks while ensuring that our work remained ours. We did consult ChatGPT for the following:

* **Syntax Assistance**: When we encountered errors related to SQL functions or timestamp operations, we asked ChatGPT for guidance on correcting syntax issues.

* **Troubleshooting Help**: For ambiguous error messages or unexpected query behavior, we consulted ChatGPT to understand potential causes and confirm our manual fixes.

Our team carefully reviewed and refined all AI-generated contributions, confirming their accuracy and originality in accordance with BU’s academic integrity and generative AI policies.


## **7. Reference**
- NYC Open Data https://data.cityofnewyork.us/Transportation/Pickups-and-Drop-offs-by-Taxi-Zone-and-Industry/c5iv-bn4s/about_data
- Google BigQuery Timestamp Functions -- https://docs.cloud.google.com/bigquery/docs/reference/standard-sql/timestamp_functions
- Combining aggregate and non-aggregate values in SQL using joins and over clause -- https://www.geeksforgeeks.org/sql/combining-aggregate-and-non-aggregate-values-in-sql-using-joins-and-over-clause/
- Green Cab -- https://www.nyc.gov/site/tlc/businesses/green-cab.page
- What's With the Green Taxis in NYC? -- https://newyorkdearest.com/green-taxis-nyc-cab/